README.md
# Dashboard_Streamlit_Pix
Dashboard para calcular quantas transações cada vendedor fez no pix


CHANGELOG
# Pequenos ajustes (15/12/23)
- Adequação do código para obedecer regras de linter
- Decorador adicionado a métodos estáticos
- Variável de ambiente de testes diferente do ambiente de produção
- Novas docstrings incluídas
- Migração do repositório para um novo, antigo mantido sob permissão privada por motivo de registro histórico

# Migração para MySQL
- Portar consultas e inserções funcionais do SQLite, SGBD original do sistema, para MySQL/MariaDB

## 08/11/23
- strptime para DATE_FORMAT: Conversão de DATETIME para String feita usando um método diferente
- Testado em ambiente MariaDB versão 11.1.2 (Docker) com Python 3.11.5 (versão atual do Arch Linux)
- Tentativa de utilizar o conector MariaDB versão 1.1.8, gerando erros ao tentar inserir dados. [Método construtor padrão do conector é Timestamp para conter datetime](https://mariadb-corporation.github.io/mariadb-connector-python/module.html#type-constructors), incompatível com a versão atual do MariaDB
- Conector alterado para mysqlclient versão 2.2.0, funcional
- Criação de banco de dados e tabelas automaticamente ao inicializar, caso não estejam presentes

tests/test_Models.py
import unittest
import logging
import os
from datetime import datetime, timedelta
import random
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
import pandas as pd
from models import Operador,Transacao, Base

TEST_DATABASE_URL = os.getenv("TEST_DATABASE_URL")
if TEST_DATABASE_URL is None:
    raise ValueError("Variável de ambiente TEST_DATABASE_URL ausente")
logging.basicConfig(level=logging.ERROR)
logger = logging.getLogger(__name__)

"""
debug do sqlalchemy
logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)
"""
class TestOperador(unittest.TestCase):
    """Testa as funcionalidades de gravação e consulta de operadores no banco de dados
    """
    @classmethod
    def setUpClass(cls):
        """Prepara a conexão ao banco de testes e cria as tabelas
        """
        logger.info("Preparando conexão ao banco")
        cls.url_banco_testes = TEST_DATABASE_URL
        cls.engine = create_engine(cls.url_banco_testes)
        logger.info("Criando tabelas de teste")
        Base.metadata.create_all(bind=cls.engine)
        Session = sessionmaker(cls.engine, autoflush=False)
        cls.session = Session()
        logger.info("Gerando dataframe operadores")
        cls.gerar_dataframe_operadores(cls)
        cls.gerar_dataframe_transacoes(cls)

    def gerar_dataframe_operadores(self):
        """Gera o dataframe com nomes e códigos de exemplo para gravar no banco
        """
        lista_operadores = []
        # nomes aleatórios, não tem intenção de representar pessoas reais
        nomes = ["Maria", "Jose", "Raimunda", "Kellen", "Francisca", 
                "Estevan", "Adriana", "Eliane", "Edilma", "Andrei"]
        logger.info("Tamanho da lista de nomes: %s", len(nomes))
        for i, nome in enumerate(nomes):
            d = {
                "Código": i+1,
                "Nome": f"{nome}",
                "Logname": f"{nome.lower()}{i+1}"
            }
            logger.info("Operador gerado: %s",d)
            lista_operadores.append(d)
        self.df_operadores = pd.DataFrame(lista_operadores)

    def gerar_dataframe_transacoes(self):
        """Com valores gerados dinâmicamente cria um dataframe 
        para inserir no banco de dados
        """
        lista_transacoes = []
        data_transacao_base = datetime.strptime(
            "01/01/2023 00:00:00", "%d/%m/%Y %H:%M:%S"
            )
        lista_possiveis_nsu = list(range(1, 1000))
        random.shuffle(lista_possiveis_nsu)

        while len(lista_transacoes) < 20:
            # Gera valores semi-aleatórios para preencher o dataframe
            
            d = {
                "Nsu": lista_possiveis_nsu.pop(),
                "Operador": random.randint(1, 10),
                "Pdv": f"P00{random.randint(1, 30)}",
                "Transacao": "Compra Pix", # if random.choice([True, False]) else "Dinheiro",
                "Estado Transacao": "Efetuada PDV", # if random.choice([True, False]) else "Negada",
                "Datetime": data_transacao_base + timedelta(minutes=len(lista_transacoes)) # soma minutos equivalente ao tamanho da lista total
            }
            lista_transacoes.append(d)
        self.df_transacoes = pd.DataFrame(lista_transacoes)

    def test_a_gravar_no_banco(self):
        """Grava os operadores gerados no banco de dados
        nomeado com "a" para execução na ordem correta
        """
        logger.info("Iniciando gravação no banco")
        Operador.gravar_banco(self.df_operadores, self.session)
        todos = Operador.todos(self.session)
        logger.info("Tamanho da lista de todos os operadores: %s", len(todos))
        self.assertEqual(len(todos),10)

    def test_b_consulta_operador(self):
        """Verifica se os operadores gerados estão presentes no banco
        """
        op_maria = Operador.operador_por_codigo(self.session, 1)
        op_andrei = Operador.operador_por_codigo(self.session, 10)
        logger.info("Consultando no banco pelas informações de operadores esperadas")
        self.assertIsNotNone(op_maria)
        self.assertIsNotNone(op_andrei)
        self.assertEqual(op_maria.nome, "Maria")
        self.assertEqual(op_andrei.nome, "Andrei")

    def test_c_gravar_transacoes_banco(self):
        """Grava as transações geradas no banco
        """
        logger.info("Grava as transacoes no banco")
        Transacao.gravar_banco(self.df_transacoes, self.session)

    def test_d_ranking(self):
        """Consulta as transações gravadas no banco com consulta válida e uma inválida
        """
        logger.info("Gerando consulta no banco com data válida")
        ranking_range_data_correto = Transacao.ranking_range_data(
            self.session, 
            datetime.strptime("01/01/2023 00:00:00", "%d/%m/%Y %H:%M:%S"),
            datetime.strptime("31/12/2023 00:00:00", "%d/%m/%Y %H:%M:%S")
            )
        self.assertIsNotNone(ranking_range_data_correto)

        logger.info("Gerando consulta no banco com data inválida")
        try:
            _ = Transacao.ranking_range_data(
                self.session, 
                datetime.strptime("01/01/1990 00:00:00", "%d/%m/%Y %H:%M:%S"),
                datetime.strptime("31/12/1990 00:00:00", "%d/%m/%Y %H:%M:%S")
                )
        except Transacao.exc.ConsultaInvalida:
            self.assertTrue(True)

    @classmethod
    def tearDownClass(cls):
        """Limpa o banco de dados para evitar resíduos de interferência no teste
        """
        cls.session.commit()
        logger.info("Teste concluído, eliminando tabelas do banco")
        Base.metadata.drop_all(bind=cls.engine)
        cls.session.close()


if __name__ == '__main__':
    unittest.main()

main.py
from models import Operador, Transacao, Base
from parsers import ParserPlanilhaOperadores, ParserRelatorioSITEF
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy import select, extract
from datetime import datetime

if __name__ == "__main__":
    engine = create_engine('sqlite:///pix.db')
    Session = sessionmaker(engine)
    session = Session()
    Base.metadata.create_all(engine)
    
    #hoje = datetime.now()
    data_inicio_str = "21/10/2023 00:00:00"
    data_inicio = datetime.strptime(data_inicio_str, "%d/%m/%Y %H:%M:%S")
    data_fim_str = "21/10/2023 23:59:59"
    data_fim = datetime.strptime(data_fim_str, "%d/%m/%Y %H:%M:%S")
    
    #df = Transacao.ranking_range_data(session,data_inicio,data_fim)
    df = Transacao.tabela_mensal_quantidade_transacoes(session, 10, 2023)
    print(df)
    
    
    #df = Operador.todos_para_df(session)
    #print(df)
    #r = Operador.todos(session)
    #for x in r:
    #    print(x.codigo_operador)
    #q = session.query(Transacao).filter(extract('month', Transacao.data_transacao)==10).all()
    #q = select(Transacao).where(Transacao.data_transacao.like("2023-10%"))
    #q = select(Transacao).where(extract("month", Transacao.data_transacao) == "10").where(extract("day", Transacao.data_transacao) == "21")
    #result = session.execute(q)
    #result.scalars()
    #for t in result.scalars().all():
    #    print(t)
    session.close()


parsers.py
import pandas as pd
import io
from datetime import datetime

class ParserRelatorioSITEF():
    def __init__(self, csv_conteudo:io.StringIO) -> None:
        self.df = pd.read_csv(csv_conteudo, sep=";")
        self._filtrar_colunas()
        self._converter_datetime()
        self._filtrar_transacoes_efetuadas()
        self.df["Operador"] = self.df["Operador"].astype(int)
        
    def _filtrar_colunas(self):
        self.df = self.df[["Data", "Hora", "Pdv", "Transacao", "Operador", "Estado Transacao", "Nsu"]]
        
    def _converter_datetime(self):
        """Gera a coluna "Datetime e converte para um objeto datetime mais versátil que string
        """
        
        self.df['Datetime'] = self.df.apply(lambda x: f"{x['Data']} {x['Hora']}", axis=1)
        self.df['Datetime'] = self.df.apply(lambda x: datetime.strptime(f"{x['Datetime']}", "%d/%m/%Y %H:%M:%S"), axis=1)
    
    def _filtrar_transacoes_efetuadas(self):
        """Descarta todas as transações que falharam ou que foram em outra forma de pagamento senão pix
        """
        self.df = self.df[self.df["Estado Transacao"] == "Efetuada PDV"]
        self.df = self.df[self.df["Transacao"] == "Compra Pix"]


class ParserPlanilhaOperadores():
    def __init__(self, planilha) -> None:
        self.df = pd.read_csv(planilha, sep=";", usecols=['Código', 'Nome', 'Logname'], encoding = 'cp1252')


    
if __name__ == "__main__":
    pass
    

dashboard.py
import io
import os
from datetime import datetime
import sqlalchemy
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
import streamlit as st
from models import Operador, Transacao, Base
from parsers import ParserPlanilhaOperadores, ParserRelatorioSITEF


DATABASE_URL = os.getenv("DATABASE_URL")
if DATABASE_URL is None:
    raise Exception("Variável de ambiente 'DATABASE_URL' faltando, necessária para iniciar")
# problemas ao gravar campos DateTime usando o conector do MariaDB, este usa Timestamp e não Datetime por padrão

engine = create_engine(DATABASE_URL)
Base.metadata.create_all(engine)
Session = sessionmaker(engine, autoflush=False)
session = Session()

def atualizar_operadores(session, conteudo_csv_operadores):
    o = ParserPlanilhaOperadores(conteudo_csv_operadores)
    Operador.gravar_banco(o.df, session)

def atualizar_transacoes(session, conteudo_csv_transacoes):
    
    t = ParserRelatorioSITEF(conteudo_csv_transacoes)
    Transacao.gravar_banco(t.df, session)
    

def gerar_ranking(session, data_inicial, data_final):
    df_ranking = Transacao.ranking_range_data(session, data_inicial, data_final)
    df_ranking["codigo_operador"] = df_ranking["codigo_operador"].astype(int)
    
    return df_ranking

hoje = datetime.now()
df = Operador.todos_para_df(session)
st.set_page_config(layout="wide")
col1, col2, col3 = st.columns(3)

#data_inicio_str = "21/10/2023 00:00:00"
#data_inicio = datetime.strptime(data_inicio_str, "%d/%m/%Y %H:%M:%S")
#data_fim_str = "21/10/2023 23:59:59"
#data_fim = datetime.strptime(data_fim_str, "%d/%m/%Y %H:%M:%S")

with col1:
   input_data_inicio = st.date_input("Data de início", value=hoje, format="DD/MM/YYYY")
   data_inicio = datetime.strptime(f"{input_data_inicio} 00:00:00", "%Y-%m-%d %H:%M:%S")
   

with col2:
    
    input_data_fim = st.date_input("Data de fim", value=hoje, format="DD/MM/YYYY")
    data_fim = datetime.strptime(f"{input_data_fim} 23:59:59", "%Y-%m-%d %H:%M:%S")
    
lateral_upload_relatorios = st.sidebar
with lateral_upload_relatorios:
    csv_operadores = st.file_uploader("Atualizar relação de operadores", type=["csv"])
    if csv_operadores is not None:
        st.button("Atualizar operadores", on_click=atualizar_operadores, args=(session, csv_operadores))
    
    csv_transacoes = st.file_uploader("Atualizar transações", type=["csv"])
    if csv_transacoes is not None:
        
        st.button("Atualizar transações", on_click=atualizar_transacoes, args=(session, csv_transacoes))


st.title('Ranking de transações do Pix por operador')

try:
    df_ranking = gerar_ranking(session, data_inicio, data_fim)
    st.dataframe(df_ranking,use_container_width=True, height=1500)
except Transacao.exc.ConsultaInvalida as e:
    "Período sem transações para exibir, verifique a data ou alimente relatório deste período"

st.title('Quantidade de transações por período')
try:
    mes = input_data_inicio.strftime("%m")
    ano = input_data_inicio.strftime("%Y")
    df_mensal = Transacao.tabela_mensal_quantidade_transacoes(
        session, 
        int(mes),
        int(ano)
        )
    st.dataframe(df_mensal, use_container_width=True)
    st.bar_chart(data=df_mensal, x="data_movimento", y="contagem_transacoes")
except Transacao.exc.ConsultaInvalida as e:
    f"Período do mês {mes}/{ano} sem transações para exibir"
#print(df_ranking)


models.py
from typing import List
from datetime import datetime
import logging
import sqlalchemy
from sqlalchemy import Integer, String, ForeignKey, DateTime
from sqlalchemy.orm import DeclarativeBase, Mapped, relationship, mapped_column
from sqlalchemy import select, between
from sqlalchemy.sql.expression import func
import pandas as pd

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class Base(DeclarativeBase):
    pass


class Operador(Base):
    """Classe compatível com SQLAlchemy que representa um operador
    """
    __tablename__ = 'operador'
    id: Mapped[int] = mapped_column(primary_key = True)
    codigo_operador:Mapped[int] = mapped_column(Integer, unique=True)
    nome:Mapped[str] = mapped_column(String(100))
    login:Mapped[str] = mapped_column(String(30))
    transacao:Mapped[List["Transacao"]] = relationship(back_populates="operador")

    def __repr__(self) -> str:
        return f"Código operador: {self.codigo_operador}, Nome: {self.nome}, Login: {self.login}"

    def to_dict(self) -> dict:
        """Gera um dicionário com os atributos da classe
        """
        return {
            "id": self.id,
            "codigo_operador": self.codigo_operador,
            "nome": self.nome,
            "login": self.login
        }

    @staticmethod
    def todos(session:sqlalchemy.orm.session.Session) -> list:
        """Retorna todos os operadores cadastrados

        Args:
            session (sqlalchemy.orm.session.Session): Sessão do SQLALchemy

        Returns:
            list: Lista com objetos Operador
        """
        r = session.execute(select(Operador)).scalars().all()
        return r
    
    @staticmethod
    def todos_para_df(session:sqlalchemy.orm.session.Session) -> list:
        """Retorna todos os operadores cadastrados na forma de um dataframe
        Usa o método to_dict() e a lista de colunas para gerar o dataframe

        Args:
            session (sqlalchemy.orm.session.Session): Sessão do SQLALchemy

        Returns:
            pd.DataFrame: DF com todos os operadores 
        """
        lista_dicts = []
        r = session.execute(select(Operador)).scalars().all()
        for x in r:
            operador_dict = x.to_dict()
            lista_dicts.append(operador_dict)
        return pd.DataFrame(lista_dicts, columns=Operador.__table__.columns.keys())
        
    @staticmethod
    def operador_por_codigo(session:sqlalchemy.orm.session.Session, cod_operador:int):
        """Consulta no banco um operador a partir do código de operador

        Args:
            session (sqlalchemy.orm.session.Session): Sessão com conexão ao banco
            cod_operador (int): Código numérico que identifica operador
        """
        result = session.execute(select(Operador).where(Operador.codigo_operador == cod_operador)).first()
        if result:
            return result[0]

    @staticmethod
    def existe(session:sqlalchemy.orm.session.Session, cod_operador:int) -> bool:
        """Verifica se operador existe no banco baseado no código identificador

        Args:
            session (sqlalchemy.orm.session.Session): Session do SQLAlchemy
            nsu (int): Identificador de operador

        Returns:
            bool: True se operador existe, False se não
        """
        result = session.execute(select(Operador).where(Operador.codigo_operador == cod_operador)).first()
        if result:
            return True
        else:
            return False

    @staticmethod
    def gravar_banco(df:pd.DataFrame, session:sqlalchemy.orm.session.Session):
        """Grava os dados de transações do dataframe no banco identificado pelo Session

        Args:
            df (pandas.DataFrame): DataFrame com dados de transações
            session (sqlalchemy.orm.session.Session): Session do SQLAlchemy
        """
        for ind in df.index:
            codigo = int(df['Código'][ind])
            nome = df['Nome'][ind]
            login = df['Logname'][ind]
            if Operador.existe(session, codigo):
                logger.info("Operador %s já existe no banco", nome)
            else:
                model_operador = Operador(
                codigo_operador=codigo,
                nome=nome,
                login=login
            )
                session.add(model_operador)
                logger.info("Gravando operador %s no banco", nome)
                del model_operador
        session.commit()

class Transacao(Base):
    """Classe compatível com SQLAlchemy que representa uma transação
    Attrs:
        nsu
        codigo_operador
        pdv
        transacao
        estado_transacao
        data_transacao
    """
    __tablename__ = 'transacao'
    id:Mapped[int] = mapped_column(primary_key = True)
    nsu:Mapped[str] = mapped_column(String(30), unique=True)
    data_transacao = mapped_column(DateTime, nullable=False)
    pdv:Mapped[int] = mapped_column(String(10))
    tipo_transacao:Mapped[str] = mapped_column(String(15))
    estado_transacao:Mapped[str] = mapped_column(String(30))
    codigo_operador:Mapped[int] = mapped_column(ForeignKey("operador.codigo_operador"))
    operador:Mapped["Operador"] = relationship(back_populates="transacao")
    
    class exc():
        class OperadorAusente(Exception):
            """Exceção para operador ausente no banco de dados
            """
            pass

        class ConsultaInvalida(Exception):
            """Exceção para consulta inválida no banco
            """
            pass

    def __repr__(self):
        return f"NSU: {self.nsu}, Data da transação: {self.data_transacao}, PDV: {self.pdv}"

    def to_dict(self):
        """Transforma em um dicionário

        Returns:
            dict: dicionário com os atributos do objeto
        """
        if self.operador is not None:
            return {
                "id": self.id,
                "nsu": self.nsu,
                "data_transacao": self.data_transacao,
                "codigo_operador": self.codigo_operador,
                "nome_operador": self.operador.nome
            }
        else:
             return {
                "id": self.id,
                "nsu": self.nsu,
                "data_transacao": self.data_transacao,
                "codigo_operador": self.codigo_operador,
                "nome_operador": "OPERADOR NÃO CADASTRADO"
            }

    @staticmethod
    def ranking_range_data(session:sqlalchemy.orm.session.Session, data_inicial:datetime, data_final:datetime) -> pd.DataFrame:
        """Consulta todas as transações em determinado período de tempo

        Args:
            session (sqlalchemy.orm.session.Session): Sessão para acesso ao banco
            data_inicial (datetime): Data de início
            data_final (datetime): Data de fim

        Raises:
            Transacao.exc.ConsultaInvalida: Período informado na consulta é inválido

        Returns:
            pd.DataFrame: DataFrame com os dados da consulta
        """
        tabela_transacao = Transacao.__table__
        tabela_operador = Operador.__table__
        consulta = select(
            tabela_transacao.c.codigo_operador,tabela_operador.c.nome, func.count().label("contagem")
            ).join(
                tabela_operador
                ).where(
                    between(Transacao.data_transacao, data_inicial, data_final)
                    ).group_by(
                tabela_transacao.c.codigo_operador
                ).order_by("contagem")
        r = session.execute(consulta).all()
        if r:
            df = pd.DataFrame(r, columns=["codigo_operador", "nome_operador", "contagem_pix"])
            df["codigo_operador"] = df["codigo_operador"].astype(int)
            return df.sort_values("contagem_pix",ascending=False)
        else:
            raise Transacao.exc.ConsultaInvalida("Ranking inválido, não há pix registrados")
    
    @staticmethod
    def tabela_mensal_quantidade_transacoes(session:sqlalchemy.orm.session.Session, mes: int, ano: int) -> pd.DataFrame:
        """
        Args:
        session (sqlalchemy.orm.session.Session)
        mes (int): formato MM
        ano (int): formato YYYY
        """
        
        
        """
        Dataframe com a seguinte query:
        SELECT strftime('%d/%m/%Y', datetime(transacao.data_transacao)) AS "Data", COUNT(*) AS "Contagem de Transações",
        strftime('%H:%M', min(datetime(transacao.data_transacao))) as "Primeira transação", strftime('%H:%M',max(datetime(transacao.data_transacao))) as "Última transação"
        FROM transacao
        WHERE strftime('%Y-%m', transacao.data_transacao) = '2023-10'
        GROUP BY strftime('%d/%m/%Y', datetime(transacao.data_transacao))
        ORDER BY datetime(transacao.data_transacao);

        """
        tabela_transacao = Transacao.__table__
        
        consulta = select(
            sqlalchemy.func.date_format(tabela_transacao.c.data_transacao,'%d/%m/%Y'), func.count().label("Contagem"), 
            sqlalchemy.func.min(sqlalchemy.func.date_format(tabela_transacao.c.data_transacao, '%H:%i')),
            sqlalchemy.func.max(sqlalchemy.func.date_format(tabela_transacao.c.data_transacao, '%H:%i')),
            ).where(sqlalchemy.func.date_format(tabela_transacao.c.data_transacao, '%Y-%m') == f"{int(ano)}-{int(mes)}"
                ).group_by(sqlalchemy.func.date_format(tabela_transacao.c.data_transacao,'%d/%m/%Y'))
        r = session.execute(consulta).all()
        
        if r:
            df = pd.DataFrame(r, columns=["data_movimento", "contagem_transacoes", "primeira_transacao", "ultima_transacao"])
            return df
        else:
            raise Transacao.exc.ConsultaInvalida("Período inválido")
    
    @staticmethod
    def existe(session:sqlalchemy.orm.session.Session, nsu:int) -> bool:
        """Verifica se transação existe no banco baseado no código identificador

        Args:
            session (sqlalchemy.orm.session.Session): Session do SQLAlchemy
            nsu (int): Identificador de transação

        Returns:
            bool: True se transação existe, False se não
        """
        result = session.execute(select(Transacao).where(Transacao.nsu == nsu)).first()
        if result:
            return True
        else:
            return False
    
    @staticmethod
    def gravar_banco(df:pd.DataFrame, session:sqlalchemy.orm.session.Session):
        """Grava os dados de operadores do dataframe no banco identificado pelo Session

        Args:
            df (pandas.DataFrame): DataFrame com dados de operador
            session (sqlalchemy.orm.session.Session): Session do SQLAlchemy
        """
        for ind in df.index:
            pdv = df['Pdv'][ind]
            data_transacao = df['Datetime'][ind]
            nsu = f"{data_transacao}-{pdv}"
            codigo_operador = int(df['Operador'][ind])
            tipo_transacao = df['Transacao'][ind]
            estado_transacao = df['Estado Transacao'][ind]
            

            if Transacao.existe(session, nsu):
                logger.info("Transacao %s já existe no banco", data_transacao)
            else:
                if Operador.existe(session, int(codigo_operador)):
                    model_transacao = Transacao(
                    nsu=nsu,
                    codigo_operador=int(codigo_operador),
                    pdv=pdv,
                    tipo_transacao=tipo_transacao,
                    estado_transacao=estado_transacao,
                    data_transacao=data_transacao
                    )   
                
                    session.add(model_transacao)
                    logger.info("gravando transacao da data %s no banco", data_transacao)
                    del model_transacao
                else:
                    raise Transacao.exc.OperadorAusente("Operador ausente, atualize a tabela de operadores e tente novamente ", codigo_operador)   
        session.commit()

if __name__ == "__main__":
    pass
    